eth/beacon/helpers.py

```diff
@@ -20,3 +170,100 @@ def get_active_validator_indices(dynasty: int,
         if (validator.start_dynasty <= dynasty and dynasty < validator.end_dynasty):
             o.append(index)
     return o
+
+
+#
+# Shuffling
+#
+def shuffle(lst: List[Any],
+            seed: Hash32) -> List[Any]:
+    lst_count = len(lst)
+    assert lst_count <= 16777216
+    o = [x for x in lst]
+    source = seed
+    i = 0
+    while i < lst_count:
+        source = blake(source)
+        for pos in range(0, 30, 3):
+            m = int.from_bytes(source[pos:pos + 3], 'big')
+            remaining = lst_count - i
+            if remaining == 0:
+                break
+            rand_max = 16777216 - 16777216 % remaining
+            if m < rand_max:
+                replacement_pos = (m % remaining) + i
+                o[i], o[replacement_pos] = o[replacement_pos], o[i]
+                i += 1
+    return o
+
+
+@to_tuple
+def split(lst: List[Any], number: int) -> List[Any]:
+    list_length = len(lst)
+    return [
+        lst[(list_length * i // number): (list_length * (i + 1) // number)]
+        for i in range(number)
+    ]
+
+
+@to_tuple
+def get_new_shuffling(seed: Hash32,
+                      validators: List['ValidatorRecord'],
+                      dynasty: int,
+                      crosslinking_start_shard: int,
+                      beacon_config: 'BeaconConfig') -> List[List[ShardAndCommittee]]:
+    cycle_length = beacon_config.cycle_length
+    min_committee_size = beacon_config.min_committee_size
+    shard_count = beacon_config.shard_count
+    avs = get_active_validator_indices(dynasty, validators)
+    if len(avs) >= cycle_length * min_committee_size:
+        committees_per_slot = min(
+            len(avs) // cycle_length // (min_committee_size * 2) + 1,
+            shard_count // cycle_length
+        )
+        slots_per_committee = 1
+    else:
+        committees_per_slot = 1
+        slots_per_committee = 1
+        while (len(avs) * slots_per_committee < cycle_length * min_committee_size and
+               slots_per_committee < cycle_length):
+            slots_per_committee *= 2
```

Comment:
It seems like we should be able to compute the final value without the `while` loop with some simple algebra.

Reply:
I reduced the problem to:
```python
while a * b < c:
    b ** 2
```

and then not sure where else can be reduced. 

There might be other way beyond trying to calcuate `_get_shuffling_committee_slot_portions()`, but I haven't figured it out...  😨

Reply:
Not 100% sure this is right, but I think it's equivalent.  You'd do `b = b ** 2 ** int(math.ceil(n))` to get the final `B` value.

![photo_2018-10-02_13-38-37](https://user-images.githubusercontent.com/824194/46372639-f844bf00-c648-11e8-82f0-f5140ccbece4.jpg)

Reply:
Bad news:
1. slap myself(!!!). a typo and misled you, it should be:
    ```python
    b = 1
    while a * b < c:
        b *= 2
    ```
    So since `b=1` initially, it's `2^N <= c/a`. -> `N <= ln(c/a)`

2. Tried `B = 2 ** int(math.log2(c / a)) when c // a > 1` and benchmarked. Unfortunately it doesn't improve much in our case. Basically because `B` is `slots_per_committee`, the edge case numbers in the spec are:
    ```
    a is active_validators_length,  domain is between [MIN_COMMITTEE_SIZE, MAX_VALIDATOR_COUNT] = [128, 4194304]    # the lower bound may be higher
    c would be CYCLE_LENGTH * MIN_COMMITTEE_SIZE, smallest: 64 * 128
    ```
    So the possible value of B is between [1, 64]. But `math` API strategy has advantage when number is large.

Many many thanks! ✨

Reply:
If you know all possible b values, what about doing a binary search across `[1, 64]`?  That should give you a loop which is guaranteed to terminate and fewer computation steps.

Reply:
Also, I don't care which method you go with at this point.  Optimization rabbit hole is deep and it's fine to just pick an approach and move on.

Reply:
> what about doing a binary search across [1, 64]? 

Good point. Will construct and benchmark it later off this PR.
I personally think it's interesting to explore the solutions. 🚂But also keep the schedule in mind.